pkgs/ffigen/lib/src/header_parser/sub_parsers/api_availability.dart

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import '../../config_provider/config_types.dart';
+import '../clang_bindings/clang_bindings.dart' as clang_types;
+import '../data.dart';
+import '../utils.dart';
+
+bool isApiAvailable(clang_types.CXCursor cursor) {
+  final api = ApiAvailability.fromCursor(cursor);
+  return api.isAvailable(config.objCMinTargetVersion);
+}
+
+class ApiAvailability {
+  final bool alwaysDeprecated;
+  final bool alwaysUnavailable;
+  PlatformAvailability? ios;
+  PlatformAvailability? macos;
+
+  ApiAvailability({
+    this.alwaysDeprecated = false,
+    this.alwaysUnavailable = false,
+    this.ios,
+    this.macos,
+  });
+
+  static ApiAvailability fromCursor(clang_types.CXCursor cursor) {
+    final platformsLength = clang.clang_getCursorPlatformAvailability(
+        cursor, nullptr, nullptr, nullptr, nullptr, nullptr, 0);
+
+    final alwaysDeprecated = calloc<Int>();
+    final alwaysUnavailable = calloc<Int>();
+    final platforms =
+        calloc<clang_types.CXPlatformAvailability>(platformsLength);
+
+    clang.clang_getCursorPlatformAvailability(cursor, alwaysDeprecated, nullptr,
+        alwaysUnavailable, nullptr, platforms, platformsLength);
+
+    PlatformAvailability? ios;
+    PlatformAvailability? macos;
+
+    for (var i = 0; i < platformsLength; ++i) {
+      final platform = platforms[i];
+      final platformAvailability = PlatformAvailability(
+        introduced: platform.Introduced.triple,
+        deprecated: platform.Deprecated.triple,
+        obsoleted: platform.Obsoleted.triple,
+        unavailable: platform.Unavailable != 0,
+      );
+      switch (platform.Platform.string()) {
+        case 'ios':
+          ios = platformAvailability;
+          break;
+        case 'macos':
+          macos = platformAvailability;
+          break;
+      }
+    }
+
+    final api = ApiAvailability(
+      alwaysDeprecated: alwaysDeprecated.value != 0,
+      alwaysUnavailable: alwaysUnavailable.value != 0,
+      ios: ios,
+      macos: macos,
+    );
+
+    for (var i = 0; i < platformsLength; ++i) {
+      clang.clang_disposeCXPlatformAvailability(platforms + i);
+    }
+    calloc.free(alwaysDeprecated);
+    calloc.free(alwaysUnavailable);
+    calloc.free(platforms);
+
+    return api;
+  }
+
+  bool isAvailable(ObjCTargetVersion minVers) {
+    // If no minimum versions are specified, everything is available.
+    if (minVers.ios == null && minVers.macos == null) {
+      return true;
+    }
+
+    if (alwaysDeprecated || alwaysUnavailable) {
+      return false;
+    }
+
+    for (final (plat, minVer) in [(ios, minVers.ios), (macos, minVers.macos)]) {
```

Comment:
I think we should try to add an availability comment or annotation if an API doesn't have the same status (available, deprecated, obsolete, unavailable) for the whole range of versions requested.